File: container-di/src/main/scala/com/yahoo/container/di/componentgraph/core/ComponentNode.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di.componentgraph.core

import java.lang.reflect.{Constructor, InvocationTargetException, Modifier, ParameterizedType, Type}
import java.util.logging.Logger

import com.google.inject.Inject
import com.yahoo.component.{AbstractComponent, ComponentId}
import com.yahoo.config.ConfigInstance
import com.yahoo.container.di.componentgraph.Provider
import com.yahoo.container.di.componentgraph.core.ComponentNode._
import com.yahoo.container.di.componentgraph.core.Node.equalEdges
import com.yahoo.container.di.{ConfigKeyT, JavaAnnotation, createKey, makeClassCovariant, preserveStackTrace, removeStackTrace}
import com.yahoo.vespa.config.ConfigKey

import scala.language.postfixOps

/**
 * @author Tony Vaagenes
 * @author gjoranv
 */
class ComponentNode(componentId: ComponentId,
                    val configId: String,
                    clazz: Class[_ <: AnyRef],
                    val XXX_key: JavaAnnotation = null) // TODO expose key, not javaAnnotation
  extends Node(componentId)
{
  require(!isAbstract(clazz), "Can't instantiate abstract class " + clazz.getName)

  var arguments : Array[AnyRef] = _

  val constructor: Constructor[AnyRef] = bestConstructor(clazz)

  var availableConfigs: Map[ConfigKeyT, ConfigInstance] = null

  override val instanceKey = createKey(clazz, XXX_key)

  override val instanceType = clazz

  override def usedComponents: List[Node] = {
    require(arguments != null, "Arguments must be set first.")
    arguments.collect{case node: Node => node}.toList
  }

  override val componentType: Class[AnyRef] = {
    def allSuperClasses(clazz: Class[_], coll : List[Class[_]]) : List[Class[_]] = {
      if (clazz == null) coll
      else allSuperClasses(clazz.getSuperclass, clazz :: coll)
    }

    def allGenericInterfaces(clazz : Class[_]) = allSuperClasses(clazz, List()) flatMap (_.getGenericInterfaces)

    def isProvider = classOf[Provider[_]].isAssignableFrom(clazz)
    def providerComponentType = (allGenericInterfaces(clazz).collect {
      case t: ParameterizedType if t.getRawType == classOf[Provider[_]] => t.getActualTypeArguments.head
    }).head

    if (isProvider) providerComponentType.asInstanceOf[Class[AnyRef]] //TODO: Test what happens if you ask for something that isn't a class, e.g. a parametrized type.
    else            clazz.asInstanceOf[Class[AnyRef]]
  }

  def setArguments(arguments: Array[AnyRef]) {
    this.arguments = arguments
  }

  def cutStackTraceAtConstructor(throwable: Throwable): Throwable = {
    def takeUntilComponentNode(elements: Array[StackTraceElement]) =
      elements.takeWhile(_.getClassName != classOf[ComponentNode].getName)

    def dropToInitAtEnd(elements: Array[StackTraceElement]) =
      elements.reverse.dropWhile(_.getMethodName != "<init>").reverse

    val modifyStackTrace = takeUntilComponentNode _ andThen dropToInitAtEnd

    val dependencyInjectorStackTraceMarker = new StackTraceElement("============= Dependency Injection =============", "newInstance", null, -1)

    if (throwable != null && !preserveStackTrace) {
      throwable.setStackTrace(modifyStackTrace(throwable.getStackTrace) :+
        dependencyInjectorStackTraceMarker)

      cutStackTraceAtConstructor(throwable.getCause)
    }
    throwable
  }

  override protected def newInstance() : AnyRef = {
    assert (arguments != null, "graph.complete must be called before retrieving instances.")

    val actualArguments = arguments.map {
      case node: Node => node.newOrCachedInstance()
      case config: ConfigKeyT => availableConfigs(config.asInstanceOf[ConfigKeyT])
      case other => other
    }

    val instance =
      try {
        constructor.newInstance(actualArguments: _*)
      } catch {
        case e: InvocationTargetException =>
          throw removeStackTrace(ErrorOrComponentConstructorException(cutStackTraceAtConstructor(e.getCause), s"Error constructing $idAndType"))
      }

    initId(instance)
  }

  private def ErrorOrComponentConstructorException(cause: Throwable, message: String) : Throwable = {
    if (cause != null && cause.isInstanceOf[Error]) // don't convert Errors to RuntimeExceptions
      new Error(message, cause)
    else
      new ComponentConstructorException(message, cause)
  }

  private def initId(component: AnyRef) = {
    def checkAndSetId(c: AbstractComponent) {
      if (c.hasInitializedId && c.getId != componentId )
        throw new IllegalStateException(
          s"Component with id '$componentId' is trying to set its component id explicitly: '${c.getId}'. " +
            "This is not allowed, so please remove any call to super() in your component's constructor.")

      c.initId(componentId)
    }

    component match {
      case component: AbstractComponent => checkAndSetId(component)
      case other => ()
    }
    component
  }

  override def equals(other: Any) = {
    other match {
      case that: ComponentNode =>
        super.equals(that) &&
          equalEdges(arguments.toList, that.arguments.toList) &&
          usedConfigs == that.usedConfigs
    }
  }

  private def usedConfigs = {
    require(availableConfigs != null, "setAvailableConfigs must be called!")
    ( arguments collect {case c: ConfigKeyT => c} map (availableConfigs) ).toList
  }

  def getAnnotatedConstructorParams: Array[(Type, Array[JavaAnnotation])] = {
    constructor.getGenericParameterTypes zip constructor.getParameterAnnotations
  }

  def setAvailableConfigs(configs: Map[ConfigKeyT, ConfigInstance]) {
    require (arguments != null, "graph.complete must be called before graph.setAvailableConfigs.")
    availableConfigs = configs
  }

  override def configKeys = {
    configParameterClasses.map(new ConfigKey(_, configId)).toSet
  }


  private def configParameterClasses: Array[Class[ConfigInstance]] = {
    constructor.getGenericParameterTypes.collect {
      case clazz: Class[_] if classOf[ConfigInstance].isAssignableFrom(clazz) => clazz.asInstanceOf[Class[ConfigInstance]]
    }
  }

  override def label = {
    val configNames = configKeys.map(_.getName + ".def").toList

    (List(instanceType.getSimpleName, Node.packageName(instanceType)) ::: configNames).
      mkString("{", "|", "}")
  }

}

object ComponentNode {
  val log = Logger.getLogger(classOf[ComponentNode].getName)

  private def bestConstructor(clazz: Class[AnyRef]) = {
    val publicConstructors = clazz.getConstructors.asInstanceOf[Array[Constructor[AnyRef]]]

    def constructorAnnotatedWithInject = {
      publicConstructors filter {_.getAnnotation(classOf[Inject]) != null} match {
        case Array() => None
        case Array(single) => Some(single)
        case _ => throwComponentConstructorException("Multiple constructors annotated with inject in class " + clazz.getName)
      }
    }

    def constructorWithMostConfigParameters = {
      def isConfigInstance(clazz: Class[_]) = classOf[ConfigInstance].isAssignableFrom(clazz)

      publicConstructors match {
        case Array() => throwComponentConstructorException("No public constructors in class " + clazz.getName)
        case Array(single) => single
        case _ =>
          log.warning("Multiple public constructors found in class %s, there should only be one. ".format(clazz.getName) +
            "If more than one public constructor is needed, the primary one must be annotated with @Inject.")
          publicConstructors.
            sortBy(_.getParameterTypes.filter(isConfigInstance).size).
            last
      }
    }

    constructorAnnotatedWithInject getOrElse constructorWithMostConfigParameters
  }

  private def throwComponentConstructorException(message: String) =
    throw removeStackTrace(new ComponentConstructorException(message))

  class ComponentConstructorException(message: String, cause: Throwable) extends RuntimeException(message, cause) {
    def this(message: String) = this(message, null)
  }

  def isAbstract(clazz: Class[_ <: AnyRef]) = Modifier.isAbstract(clazz.getModifiers)
}

File: container-di/src/main/java/com/yahoo/container/di/componentgraph/core/ComponentNode.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di.componentgraph.core;

import com.google.inject.Inject;
import com.google.inject.Key;
import com.yahoo.collections.Pair;
import com.yahoo.component.AbstractComponent;
import com.yahoo.component.ComponentId;
import com.yahoo.config.ConfigInstance;
import com.yahoo.container.di.componentgraph.Provider;
import com.yahoo.vespa.config.ConfigKey;

import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Modifier;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.logging.Logger;
import java.util.stream.Collectors;

import static com.yahoo.container.di.componentgraph.core.Exceptions.cutStackTraceAtConstructor;
import static com.yahoo.container.di.componentgraph.core.Exceptions.removeStackTrace;
import static com.yahoo.container.di.componentgraph.core.Keys.createKey;

/**
 * @author Tony Vaagenes
 * @author gjoranv
 * @author ollivir
 */
public class ComponentNode extends Node {
    private static final Logger log = Logger.getLogger(ComponentNode.class.getName());

    private final Class<?> clazz;
    private final Annotation key;
    private Object[] arguments = null;
    private final String configId;

    private final Constructor<?> constructor;

    private Map<ConfigKey<ConfigInstance>, ConfigInstance> availableConfigs = null;


    public ComponentNode(ComponentId componentId,
                         String configId,
                         Class<?> clazz, Annotation XXX_key) // TODO expose key, not javaAnnotation
    {
        super(componentId);
        if (isAbstract(clazz)) {
            throw new IllegalArgumentException("Can't instantiate abstract class " + clazz.getName());
        }
        this.configId = configId;
        this.clazz = clazz;
        this.key = XXX_key;
        this.constructor = bestConstructor(clazz);
    }

    public ComponentNode(ComponentId componentId, String configId, Class<?> clazz) {
        this(componentId, configId, clazz, null);
    }

    public String configId() {
        return configId;
    }

    @Override
    public Key<?> instanceKey() {
        return createKey(clazz, key);
    }

    @Override
    public Class<?> instanceType() {
        return clazz;
    }

    @Override
    public List<Node> usedComponents() {
        if (arguments == null) {
            throw new IllegalStateException("Arguments must be set first.");
        }
        List<Node> ret = new ArrayList<>();
        for (Object arg : arguments) {
            if (arg instanceof Node) {
                ret.add((Node) arg);
            }
        }
        return ret;
    }

    private static List<Class<?>> allSuperClasses(Class<?> clazz) {
        List<Class<?>> ret = new ArrayList<>();
        while (clazz != null) {
            ret.add(clazz);
            clazz = clazz.getSuperclass();
        }
        return ret;
    }

    @Override
    public Class<?> componentType() {
        if (Provider.class.isAssignableFrom(clazz)) {
            //TODO: Test what happens if you ask for something that isn't a class, e.g. a parameterized type.

            List<Type> allGenericInterfaces = allSuperClasses(clazz).stream().flatMap(c -> Arrays.stream(c.getGenericInterfaces())).collect(Collectors.toList());
            for (Type t : allGenericInterfaces) {
                if (t instanceof ParameterizedType && ((ParameterizedType) t).getRawType().equals(Provider.class)) {
                    Type[] typeArgs = ((ParameterizedType) t).getActualTypeArguments();
                    if (typeArgs != null && typeArgs.length > 0) {
                        return (Class<?>) typeArgs[0];
                    }
                }
            }
            throw new IllegalStateException("Component type cannot be resolved");
        } else {
            return clazz;
        }
    }

    public void setArguments(Object[] arguments) {
        this.arguments = arguments;
    }

    @Override
    protected Object newInstance() {
        if (arguments == null) {
            throw new IllegalStateException("graph.complete must be called before retrieving instances.");
        }

        List<Object> actualArguments = new ArrayList<>();
        for (Object ob : arguments) {
            if (ob instanceof Node) {
                actualArguments.add(((Node) ob).newOrCachedInstance());
            } else if (ob instanceof ConfigKey) {
                actualArguments.add(availableConfigs.get(ob));
            } else {
                actualArguments.add(ob);
            }
        }

        Object instance;
        try {
            instance = constructor.newInstance(actualArguments.toArray());
        } catch (InvocationTargetException | InstantiationException | IllegalAccessException e) {
            StackTraceElement dependencyInjectorMarker = new StackTraceElement("============= Dependency Injection =============", "newInstance", null, -1);

            throw removeStackTrace(new ComponentConstructorException("Error constructing " + idAndType(), cutStackTraceAtConstructor(e.getCause(), dependencyInjectorMarker)));
        }

        return initId(instance);
    }

    private Object initId(Object component) {
        if (component instanceof AbstractComponent) {
            AbstractComponent abstractComponent = (AbstractComponent) component;
            if (abstractComponent.hasInitializedId() && !abstractComponent.getId().equals(componentId())) {
                throw new IllegalStateException(
                        "Component with id '" + componentId() + "' is trying to set its component id explicitly: '" + abstractComponent.getId() + "'. " +
                                "This is not allowed, so please remove any call to super() in your component's constructor.");
            }
            abstractComponent.initId(componentId());
        }
        return component;
    }


    @Override
    public int hashCode() {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + Arrays.hashCode(arguments);
        result = prime * result + ((availableConfigs == null) ? 0 : availableConfigs.hashCode());
        result = prime * result + ((configId == null) ? 0 : configId.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof ComponentNode) {
            ComponentNode that = (ComponentNode) other;
            return super.equals(that) && equalEdges(Arrays.asList(this.arguments), Arrays.asList(that.arguments)) && this.usedConfigs().equals(that.usedConfigs());
        } else {
            return false;
        }
    }

    private List<ConfigInstance> usedConfigs() {
        if (availableConfigs == null) {
            throw new IllegalStateException("setAvailableConfigs must be called!");
        }
        List<ConfigInstance> ret = new ArrayList<>();
        for (Object arg : arguments) {
            if (arg instanceof ConfigKey) {
                ret.add(availableConfigs.get(arg));
            }
        }
        return ret;
    }

    protected List<Pair<Type, List<Annotation>>> getAnnotatedConstructorParams() {
        Type[] types = constructor.getGenericParameterTypes();
        Annotation[][] annotations = constructor.getParameterAnnotations();

        List<Pair<Type, List<Annotation>>> ret = new ArrayList<>();

        for (int i = 0; i < types.length; i++) {
            ret.add(new Pair<>(types[i], Arrays.asList(annotations[i])));
        }
        return ret;
    }

    public void setAvailableConfigs(Map<ConfigKey<ConfigInstance>, ConfigInstance> configs) {
        if (arguments == null) {
            throw new IllegalStateException("graph.complete must be called before graph.setAvailableConfigs.");
        }
        this.availableConfigs = configs;
    }

    @Override
    public Set<ConfigKey<ConfigInstance>> configKeys() {
        return configParameterClasses().stream().map(par -> new ConfigKey<>(par, configId)).collect(Collectors.toSet());
    }

    @SuppressWarnings("unchecked")
    private List<Class<ConfigInstance>> configParameterClasses() {
        List<Class<ConfigInstance>> ret = new ArrayList<>();
        for (Type type : constructor.getGenericParameterTypes()) {
            if (type instanceof Class && ConfigInstance.class.isAssignableFrom((Class<?>) type)) {
                ret.add((Class<ConfigInstance>) type);
            }
        }
        return ret;
    }

    @Override
    public String label() {
        LinkedList<String> configNames = configKeys().stream().map(k -> k.getName() + ".def").collect(Collectors.toCollection(LinkedList::new));

        configNames.addFirst(instanceType().getSimpleName());
        configNames.addFirst(Node.packageName(instanceType()));

        return "{" + String.join("|", configNames) + "}";
    }

    private static Constructor<?> bestConstructor(Class<?> clazz) {
        Constructor<?>[] publicConstructors = clazz.getConstructors();

        Constructor<?> annotated = null;
        for (Constructor<?> ctor : publicConstructors) {
            Annotation annotation = ctor.getAnnotation(Inject.class);
            if (annotation != null) {
                if (annotated == null) {
                    annotated = ctor;
                } else {
                    throw componentConstructorException("Multiple constructor annotated with @Inject in class " + clazz.getName());
                }
            }
        }
        if (annotated != null) {
            return annotated;
        }

        if (publicConstructors.length == 0) {
            throw componentConstructorException("No public constructors in class " + clazz.getName());
        } else if (publicConstructors.length == 1) {
            return publicConstructors[0];
        } else {
            log.warning(String.format("Multiple public constructors found in class %s, there should only be one. "
                    + "If more than one public constructor is needed, the primary one must be annotated with @Inject.", clazz.getName()));
            List<Pair<Constructor<?>, Integer>> withParameterCount = new ArrayList<>();
            for (Constructor<?> ctor : publicConstructors) {
                long count = Arrays.stream(ctor.getParameterTypes()).filter(ConfigInstance.class::isAssignableFrom).count();
                withParameterCount.add(new Pair<>(ctor, (int) count));
            }
            withParameterCount.sort(Comparator.comparingInt(Pair::getSecond));
            return withParameterCount.get(withParameterCount.size() - 1).getFirst();
        }
    }

    private static ComponentConstructorException componentConstructorException(String message) {
        return removeStackTrace(new ComponentConstructorException(message));
    }

    public static class ComponentConstructorException extends RuntimeException {
        ComponentConstructorException(String message) {
            super(message);
        }

        ComponentConstructorException(String message, Throwable cause) {
            super(message, cause);
        }
    }


    private static boolean isAbstract(Class<?> clazz) {
        return Modifier.isAbstract(clazz.getModifiers());
    }
}